Cannot use "managed" command without signing in to 1Password anymore

## What exactly are you trying to do?

With chezmoi 1.x I used a shell function to edit my dotfiles and apply them immediately. The function first lists all managed files using `chezmoi managed`, then it edits the chosen file with `chezmoi edit --apply`. This is what it looks like (it's a fish script):

```fish
function d --description 'Edit dotfiles'
    set -l file (chezmoi managed --include files | fzf --select-1 --query "$argv")
    and chezmoi edit --apply --verbose ~/"$file"
end
```

Now when I run `chezmoi managed` with chezmoi 2.x, it tries to execute the templates, some of which contain calls to `onepassword`. So this requires me to be signed in to 1Password only to get a list of the managed files.

Also once I am signed in to 1Password it takes significantly longer to list all files, presumably because it now parses all templates.

So I was wondering if there was a way to work around this, or if this is something that should be fixed in the `managed` command of chezmoi 2.x.

## What have you tried so far?

n/a

## Where else have you checked for solutions?

* [x] I have read [chezmoi's how-to guide](https://github.com/twpayne/chezmoi/blob/master/docs/HOWTO.md), and not found the answer.
* [x] I have read [chezmoi's FAQ](https://github.com/twpayne/chezmoi/blob/master/docs/FAQ.md), and not found the answer.
* [x] I have searched [chezmoi's reference guide](https://github.com/twpayne/chezmoi/blob/master/docs/REFERENCE.md), and not found the answer.

## Output of any commands you've tried with `--verbose`

```
$ chezmoi managed --verbose
chezmoi: template: private_dot_config/nvim/coc-settings.json.tmpl:2:50: executing "private_dot_config/nvim/coc-settings.json.tmpl" at <onepassword "zyuqdsryv5dyzenrn6kadrx6ve">: error calling onepassword: op get item zyuqdsryv5dyzenrn6kadrx6ve: exit status 1: [ERROR] 2021/03/28 14:36:02 You are not currently signed in. Please run `op signin --help` for instructions
```

## Output of `chezmoi doctor`

```
$ chezmoi doctor
RESULT   CHECK                MESSAGE
ok       version              v2.0.4, commit 9cc8184cc466a87c44006218e02365046e5d5f8b, built at 2021-03-24T20:20:46Z, built by homebrew
ok       os-arch              darwin/arm64
ok       config-file          /Users/adriaan/.config/chezmoi/chezmoi.json is a file
ok       source-dir           /Users/adriaan/.local/share/chezmoi is a directory
ok       suspicious-entries   no suspicious entries found in /Users/adriaan/.local/share/chezmoi
ok       dest-dir             /Users/adriaan is a directory
ok       shell                found /bin/zsh
ok       editor               found /opt/homebrew/bin/nvim
ok       git-cli              found /usr/bin/git, version 2.24.3
ok       merge-cli            found /usr/bin/vimdiff
info     age-cli              age not found in $PATH
ok       gnupg-cli            gpg not found in $PATH
ok       1password-cli        found /usr/local/bin/op, version 1.8.0
info     bitwarden-cli        bw not found in $PATH
info     gopass-cli           gopass not found in $PATH
info     keepassxc-cli        keepassxc-cli not found in $PATH
info     keepassxc-db         not set
info     lastpass-cli         lpass not found in $PATH
info     pass-cli             pass not found in $PATH
info     vault-cli            vault not found in $PATH
info     secret-cli           not set
```

--- a/cmd/managedcmd_test.go
+++ b/cmd/managedcmd_test.go
@@ -1,0 +1,87 @@
+package cmd
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/twpayne/go-vfs/v2"
+
+	"github.com/twpayne/chezmoi/internal/chezmoitest"
+)
+
+func TestManagedCmd(t *testing.T) {
+	for _, tc := range []struct {
+		name           string
+		root           interface{}
+		args           []string
+		expectedOutput string
+	}{
+		{
+			name: "simple",
+			root: map[string]interface{}{
+				"/home/user/.local/share/chezmoi/dot_file": "# contents of .file\n",
+			},
+			expectedOutput: chezmoitest.JoinLines(
+				".file",
+			),
+		},
+		{
+			name: "template",
+			root: map[string]interface{}{
+				"/home/user/.local/share/chezmoi/dot_template.tmpl": "{{ fail \"Template should not be executed\" }}\n",
+			},
+			expectedOutput: chezmoitest.JoinLines(
+				".template",
+			),
+		},
+		{
+			name: "create_template",
+			root: map[string]interface{}{
+				"/home/user/.local/share/chezmoi/create_dot_file.tmpl": "{{ fail \"Template should not be executed\" }}\n",
+			},
+			expectedOutput: chezmoitest.JoinLines(
+				".file",
+			),
+		},
+		{
+			name: "modify_template",
+			root: map[string]interface{}{
+				"/home/user/.local/share/chezmoi/modify_dot_file.tmpl": "{{ fail \"Template should not be executed\" }}\n",
+			},
+			expectedOutput: chezmoitest.JoinLines(
+				".file",
+			),
+		},
+		{
+			name: "script_template",
+			root: map[string]interface{}{
+				"/home/user/.local/share/chezmoi/run_script.tmpl": "{{ fail \"Template should not be executed\" }}\n",
+			},
+			args: []string{
+				"--include", "scripts",
+			},
+			expectedOutput: chezmoitest.JoinLines(
+				"script",
+			),
+		},
+		{
+			name: "symlink_template",
+			root: map[string]interface{}{
+				"/home/user/.local/share/chezmoi/symlink_dot_symlink.tmpl": "{{ fail \"Template should not be executed\" }}\n",
+			},
+			expectedOutput: chezmoitest.JoinLines(
+				".symlink",
+			),
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			chezmoitest.WithTestFS(t, tc.root, func(fs vfs.FS) {
+				stdout := &bytes.Buffer{}
+				require.NoError(t, newTestConfig(t, fs, withStdout(stdout)).execute(append([]string{"managed"}, tc.args...)))
+				assert.Equal(t, tc.expectedOutput, stdout.String())
+			})
+		})
+	}
+}

--- a/internal/chezmoi/actualstateentry.go
+++ b/internal/chezmoi/actualstateentry.go
@@ -56,11 +56,9 @@ func NewActualStateEntry(system System, absPath AbsPath, info os.FileInfo, err e
 		return &ActualStateFile{
 			absPath: absPath,
 			perm:    info.Mode() & os.ModePerm,
-			lazyContents: &lazyContents{
-				contentsFunc: func() ([]byte, error) {
-					return system.ReadFile(absPath)
-				},
-			},
+			lazyContents: newLazyContentsFunc(func() ([]byte, error) {
+				return system.ReadFile(absPath)
+			}),
 		}, nil
 	case os.ModeDir:
 		return &ActualStateDir{
@@ -70,15 +68,9 @@ func NewActualStateEntry(system System, absPath AbsPath, info os.FileInfo, err e
 	case os.ModeSymlink:
 		return &ActualStateSymlink{
 			absPath: absPath,
-			lazyLinkname: &lazyLinkname{
-				linknameFunc: func() (string, error) {
-					linkname, err := system.Readlink(absPath)
-					if err != nil {
-						return "", err
-					}
-					return linkname, nil
-				},
-			},
+			lazyLinkname: newLazyLinknameFunc(func() (string, error) {
+				return system.Readlink(absPath)
+			}),
 		}, nil
 	default:
 		return nil, &errUnsupportedFileType{

--- a/internal/chezmoi/lazy.go
+++ b/internal/chezmoi/lazy.go
@@ -31,6 +31,13 @@ func newLazyContents(contents []byte) *lazyContents {
 	}
 }
 
+// newLazyContentsFunc returns a new lazyContents with contentsFunc.
+func newLazyContentsFunc(contentsFunc func() ([]byte, error)) *lazyContents {
+	return &lazyContents{
+		contentsFunc: contentsFunc,
+	}
+}
+
 // Contents returns lc's contents.
 func (lc *lazyContents) Contents() ([]byte, error) {
 	if lc == nil {
@@ -65,6 +72,13 @@ func (lc *lazyContents) ContentsSHA256() ([]byte, error) {
 func newLazyLinkname(linkname string) *lazyLinkname {
 	return &lazyLinkname{
 		linkname: linkname,
+	}
+}
+
+// newLazyLinknameFunc returns a new lazyLinkname with linknameFunc.
+func newLazyLinknameFunc(linknameFunc func() (string, error)) *lazyLinkname {
+	return &lazyLinkname{
+		linknameFunc: linknameFunc,
 	}
 }
 

--- a/internal/chezmoi/sourcestate.go
+++ b/internal/chezmoi/sourcestate.go
@@ -839,26 +839,28 @@ func (s *SourceState) newSourceStateFile(sourceRelPath SourceRelPath, fileAttr F
 			}
 			return &TargetStateFile{
 				lazyContents: newLazyContents(contents),
+				empty:        true,
 				perm:         fileAttr.perm(),
 			}, nil
 		}
 	case SourceFileTypeFile:
 		targetStateEntryFunc = func(destSystem System, destAbsPath AbsPath) (TargetStateEntry, error) {
-			contents, err := sourceLazyContents.Contents()
-			if err != nil {
-				return nil, err
-			}
-			if fileAttr.Template {
-				contents, err = s.ExecuteTemplateData(sourceRelPath.String(), contents)
+			contentsFunc := func() ([]byte, error) {
+				contents, err := sourceLazyContents.Contents()
 				if err != nil {
 					return nil, err
 				}
-			}
-			if !fileAttr.Empty && isEmpty(contents) {
-				return &TargetStateRemove{}, nil
+				if fileAttr.Template {
+					contents, err = s.ExecuteTemplateData(sourceRelPath.String(), contents)
+					if err != nil {
+						return nil, err
+					}
+				}
+				return contents, nil
 			}
 			return &TargetStateFile{
-				lazyContents: newLazyContents(contents),
+				lazyContents: newLazyContentsFunc(contentsFunc),
+				empty:        fileAttr.Empty,
 				perm:         fileAttr.perm(),
 			}, nil
 		}
@@ -918,47 +920,48 @@ func (s *SourceState) newSourceStateFile(sourceRelPath SourceRelPath, fileAttr F
 				return destSystem.IdempotentCmdOutput(cmd)
 			}
 			return &TargetStateFile{
-				lazyContents: &lazyContents{
-					contentsFunc: contentsFunc,
-				},
-				perm: fileAttr.perm(),
+				lazyContents: newLazyContentsFunc(contentsFunc),
+				perm:         fileAttr.perm(),
 			}, nil
 		}
 	case SourceFileTypeScript:
 		targetStateEntryFunc = func(destSystem System, destAbsPath AbsPath) (TargetStateEntry, error) {
-			contents, err := sourceLazyContents.Contents()
-			if err != nil {
-				return nil, err
-			}
-			if fileAttr.Template {
-				contents, err = s.ExecuteTemplateData(sourceRelPath.String(), contents)
+			contentsFunc := func() ([]byte, error) {
+				contents, err := sourceLazyContents.Contents()
 				if err != nil {
 					return nil, err
 				}
+				if fileAttr.Template {
+					contents, err = s.ExecuteTemplateData(sourceRelPath.String(), contents)
+					if err != nil {
+						return nil, err
+					}
+				}
+				return contents, nil
 			}
 			return &TargetStateScript{
-				lazyContents: newLazyContents(contents),
+				lazyContents: newLazyContentsFunc(contentsFunc),
 				name:         targetRelPath,
 				once:         fileAttr.Once,
 			}, nil
 		}
 	case SourceFileTypeSymlink:
 		targetStateEntryFunc = func(destSystem System, destAbsPath AbsPath) (TargetStateEntry, error) {
-			linknameBytes, err := sourceLazyContents.Contents()
-			if err != nil {
-				return nil, err
-			}
-			if fileAttr.Template {
-				linknameBytes, err = s.ExecuteTemplateData(sourceRelPath.String(), linknameBytes)
+			linknameFunc := func() (string, error) {
+				linknameBytes, err := sourceLazyContents.Contents()
 				if err != nil {
-					return nil, err
+					return "", err
 				}
-			}
-			if isEmpty(linknameBytes) {
-				return &TargetStateRemove{}, nil
+				if fileAttr.Template {
+					linknameBytes, err = s.ExecuteTemplateData(sourceRelPath.String(), linknameBytes)
+					if err != nil {
+						return "", err
+					}
+				}
+				return string(bytes.TrimSpace(linknameBytes)), nil
 			}
 			return &TargetStateSymlink{
-				lazyLinkname: newLazyLinkname(string(bytes.TrimSpace(linknameBytes))),
+				lazyLinkname: newLazyLinknameFunc(linknameFunc),
 			}, nil
 		}
 	default:
@@ -1022,9 +1025,7 @@ func (s *SourceState) sourceStateEntry(actualStateEntry ActualStateEntry, destAb
 				return nil, err
 			}
 		}
-		lazyContents := &lazyContents{
-			contents: contents,
-		}
+		lazyContents := newLazyContents(contents)
 		sourceRelPath := parentSourceRelPath.Join(NewSourceRelPath(RelPath(fileAttr.SourceName(s.encryption.EncryptedSuffix()))))
 		return &SourceStateFile{
 			Attr:          fileAttr,
@@ -1032,6 +1033,7 @@ func (s *SourceState) sourceStateEntry(actualStateEntry ActualStateEntry, destAb
 			lazyContents:  lazyContents,
 			targetStateEntry: &TargetStateFile{
 				lazyContents: lazyContents,
+				empty:        options.Empty,
 				perm:         0o666,
 			},
 		}, nil
@@ -1050,9 +1052,7 @@ func (s *SourceState) sourceStateEntry(actualStateEntry ActualStateEntry, destAb
 			contents = autoTemplate(contents, s.TemplateData())
 		}
 		contents = append(contents, '\n')
-		lazyContents := &lazyContents{
-			contents: contents,
-		}
+		lazyContents := newLazyContents(contents)
 		sourceRelPath := parentSourceRelPath.Join(NewSourceRelPath(RelPath(fileAttr.SourceName(s.encryption.EncryptedSuffix()))))
 		return &SourceStateFile{
 			Attr:          fileAttr,

--- a/internal/chezmoi/sourcestate_test.go
+++ b/internal/chezmoi/sourcestate_test.go
@@ -931,14 +931,10 @@ func TestSourceStateRead(t *testing.T) {
 							TargetName: "file",
 							Type:       SourceFileTypeFile,
 						},
-						lazyContents: &lazyContents{
-							contents: []byte("# contents of .dir/file\n"),
-						},
+						lazyContents: newLazyContents([]byte("# contents of .dir/file\n")),
 						targetStateEntry: &TargetStateFile{
-							perm: 0o666,
-							lazyContents: &lazyContents{
-								contents: []byte("# contents of .dir/file\n"),
-							},
+							perm:         0o666,
+							lazyContents: newLazyContents([]byte("# contents of .dir/file\n")),
 						},
 					},
 				}),
@@ -1008,14 +1004,10 @@ func TestSourceStateRead(t *testing.T) {
 							TargetName: "file1",
 							Type:       SourceFileTypeFile,
 						},
-						lazyContents: &lazyContents{
-							contents: []byte("# contents of dir/file1\n"),
-						},
+						lazyContents: newLazyContents([]byte("# contents of dir/file1\n")),
 						targetStateEntry: &TargetStateFile{
-							perm: 0o666,
-							lazyContents: &lazyContents{
-								contents: []byte("# contents of dir/file1\n"),
-							},
+							perm:         0o666,
+							lazyContents: newLazyContents([]byte("# contents of dir/file1\n")),
 						},
 					},
 					"dir/file2": &SourceStateRemove{

--- a/internal/chezmoi/targetstateentry.go
+++ b/internal/chezmoi/targetstateentry.go
@@ -24,7 +24,8 @@ type TargetStateDir struct {
 // A TargetStateFile represents the state of a file in the target state.
 type TargetStateFile struct {
 	*lazyContents
-	perm os.FileMode
+	empty bool
+	perm  os.FileMode
 }
 
 // A TargetStateRemove represents the absence of an entry in the target state.
@@ -89,6 +90,16 @@ func (t *TargetStateDir) SkipApply(persistentState PersistentState) (bool, error
 
 // Apply updates actualStateEntry to match t.
 func (t *TargetStateFile) Apply(system System, persistentState PersistentState, actualStateEntry ActualStateEntry, umask os.FileMode) (bool, error) {
+	contents, err := t.Contents()
+	if err != nil {
+		return false, err
+	}
+	if !t.empty && isEmpty(contents) {
+		if _, ok := actualStateEntry.(*ActualStateAbsent); ok {
+			return false, nil
+		}
+		return true, system.RemoveAll(actualStateEntry.Path())
+	}
 	if actualStateFile, ok := actualStateEntry.(*ActualStateFile); ok {
 		// Compare file contents using only their SHA256 sums. This is so that
 		// we can compare last-written states without storing the full contents
@@ -110,10 +121,6 @@ func (t *TargetStateFile) Apply(system System, persistentState PersistentState, 
 	} else if err := actualStateEntry.Remove(system); err != nil {
 		return false, err
 	}
-	contents, err := t.Contents()
-	if err != nil {
-		return false, err
-	}
 	return true, system.WriteFile(actualStateEntry.Path(), contents, t.perm&^umask)
 }
 
@@ -122,6 +129,11 @@ func (t *TargetStateFile) EntryState(umask os.FileMode) (*EntryState, error) {
 	contents, err := t.Contents()
 	if err != nil {
 		return nil, err
+	}
+	if !t.empty && isEmpty(contents) {
+		return &EntryState{
+			Type: EntryStateTypeRemove,
+		}, nil
 	}
 	contentsSHA256, err := t.ContentsSHA256()
 	if err != nil {
@@ -242,6 +254,16 @@ func (t *TargetStateScript) SkipApply(persistentState PersistentState) (bool, er
 
 // Apply updates actualStateEntry to match t.
 func (t *TargetStateSymlink) Apply(system System, persistentState PersistentState, actualStateEntry ActualStateEntry, umask os.FileMode) (bool, error) {
+	linkname, err := t.Linkname()
+	if err != nil {
+		return false, err
+	}
+	if linkname == "" {
+		if _, ok := actualStateEntry.(*ActualStateAbsent); ok {
+			return false, nil
+		}
+		return true, system.RemoveAll(actualStateEntry.Path())
+	}
 	if actualStateSymlink, ok := actualStateEntry.(*ActualStateSymlink); ok {
 		actualLinkname, err := actualStateSymlink.Linkname()
 		if err != nil {
@@ -255,10 +277,6 @@ func (t *TargetStateSymlink) Apply(system System, persistentState PersistentStat
 			return false, nil
 		}
 	}
-	linkname, err := t.Linkname()
-	if err != nil {
-		return false, err
-	}
 	if err := actualStateEntry.Remove(system); err != nil {
 		return false, err
 	}
@@ -270,6 +288,11 @@ func (t *TargetStateSymlink) EntryState(umask os.FileMode) (*EntryState, error) 
 	linkname, err := t.Linkname()
 	if err != nil {
 		return nil, err
+	}
+	if linkname == "" {
+		return &EntryState{
+			Type: EntryStateTypeRemove,
+		}, nil
 	}
 	linknameSHA256, err := t.LinknameSHA256()
 	if err != nil {

--- a/internal/chezmoi/targetstateentry_test.go
+++ b/internal/chezmoi/targetstateentry_test.go
@@ -14,31 +14,26 @@ import (
 	"github.com/twpayne/chezmoi/internal/chezmoitest"
 )
 
-func TestTargetStateEntryApplyAndEqual(t *testing.T) {
+func TestTargetStateEntryApply(t *testing.T) {
 	targetStates := map[string]TargetStateEntry{
 		"dir": &TargetStateDir{
 			perm: 0o777 &^ chezmoitest.Umask,
 		},
 		"file": &TargetStateFile{
-			perm: 0o666 &^ chezmoitest.Umask,
-			lazyContents: &lazyContents{
-				contents: []byte("# contents of file"),
-			},
+			perm:         0o666 &^ chezmoitest.Umask,
+			lazyContents: newLazyContents([]byte("# contents of file")),
 		},
 		"file_empty": &TargetStateFile{
-			perm: 0o666 &^ chezmoitest.Umask,
+			perm:  0o666 &^ chezmoitest.Umask,
+			empty: true,
 		},
 		"file_executable": &TargetStateFile{
-			perm: 0o777 &^ chezmoitest.Umask,
-			lazyContents: &lazyContents{
-				contents: []byte("#!/bin/sh\n"),
-			},
+			perm:         0o777 &^ chezmoitest.Umask,
+			lazyContents: newLazyContents([]byte("#!/bin/sh\n")),
 		},
 		"remove": &TargetStateRemove{},
 		"symlink": &TargetStateSymlink{
-			lazyLinkname: &lazyLinkname{
-				linkname: "target",
-			},
+			lazyLinkname: newLazyLinkname("target"),
 		},
 	}
 

--- a/testdata/scripts/managed.txt
+++ b/testdata/scripts/managed.txt
@@ -1,25 +1,32 @@
 mksourcedir
 
+# test chezmoi managed
 chezmoi managed
-cmpenv stdout golden/managed
+cmp stdout golden/managed
 
+# test chezmoi managed --include=all
 chezmoi managed --include=all
-cmpenv stdout golden/managed-all
+cmp stdout golden/managed-all
 
+# test chezmoi managed --include=remove
 chezmoi managed --include=remove
-cmpenv stdout golden/managed-remove
+cmp stdout golden/managed-remove
 
+# test chezmoi managed --include=dirs
 chezmoi managed --include=dirs
-cmpenv stdout golden/managed-dirs
+cmp stdout golden/managed-dirs
 
+# test chezmoi managed --include=files
 chezmoi managed --include=files
-cmpenv stdout golden/managed-files
+cmp stdout golden/managed-files
 
+# test chezmoi managed --include=symlinks
 chezmoi managed --include=symlinks
-cmpenv stdout golden/managed-symlinks
+cmp stdout golden/managed-symlinks
 
+# test chezmoi managed --exclude=files
 chezmoi managed --exclude=files
-cmpenv stdout golden/managed-except-files
+cmp stdout golden/managed-except-files
 
 -- golden/managed --
 .create

--- a/testdata/scripts/managed.txt
+++ b/testdata/scripts/managed.txt
@@ -8,10 +8,6 @@ cmp stdout golden/managed
 chezmoi managed --include=all
 cmp stdout golden/managed-all
 
-# test chezmoi managed --include=remove
-chezmoi managed --include=remove
-cmp stdout golden/managed-remove
-
 # test chezmoi managed --include=dirs
 chezmoi managed --include=dirs
 cmp stdout golden/managed-dirs
@@ -28,6 +24,11 @@ cmp stdout golden/managed-symlinks
 chezmoi managed --exclude=files
 cmp stdout golden/managed-except-files
 
+# test that chezmoi managed does not evaluate templates
+chhome home2/user
+chezmoi managed --include=all
+cmp stdout golden/managed2
+
 -- golden/managed --
 .create
 .dir
@@ -38,6 +39,7 @@ cmp stdout golden/managed-except-files
 .executable
 .file
 .private
+.remove
 .symlink
 .template
 -- golden/managed-all --
@@ -64,12 +66,29 @@ cmp stdout golden/managed-except-files
 .executable
 .file
 .private
-.template
--- golden/managed-remove --
 .remove
+.template
 -- golden/managed-symlinks --
 .symlink
 -- golden/managed-except-files --
 .dir
 .dir/subdir
 .symlink
+-- golden/managed2 --
+.create
+.file
+.symlink
+.template
+script
+-- home/user/.local/share/chezmoi/.chezmoiremove --
+.remove
+-- home2/user/.local/share/chezmoi/create_dot_create.tmpl --
+{{ fail "Template should not be executed }}
+-- home2/user/.local/share/chezmoi/dot_template.tmpl --
+{{ fail "Template should not be executed }}
+-- home2/user/.local/share/chezmoi/modify_dot_file.tmpl --
+{{ fail "Template should not be executed }}
+-- home2/user/.local/share/chezmoi/symlink_dot_symlink.tmpl --
+{{ fail "Template should not be executed }}
+-- home2/user/.local/share/chezmoi/run_script.tmpl --
+{{ fail "Template should not be executed }}

